packages/lib/src/subscriptions.ts

```diff
@@ -0,0 +1,337 @@
+			if (!entry) {
+				throw new Error(`Unregistered Event class ${json[0]}`);
+			} else {
+				return new SubscriptionResponse(entry.eventFromJSON(json[1]));
+			}
+		} else {
+			return new SubscriptionResponse();
+		}
+	}
+}
+
+/**
+ * A subscription request sent by a subscriber, this updates what events the subscriber will be sent
+ * The permission for this request copies the permission from the event being subscribed to
+ * allChannels: false, channels: [] will unsubscribe the subscriber from all notifications
+ */
+export class SubscriptionRequest {
+	declare ["constructor"]: typeof SubscriptionRequest;
+	static type = "request" as const;
+	static src =  ["control", "instance"] as const;
+	static dst = "controller" as const;
+	static Response = SubscriptionResponse;
+	static permission(user: User, message: MessageRequest) {
+		if (typeof message.data === "object" && message.data !== null) {
+			const data = message.data as Static<typeof SubscriptionRequest.jsonSchema>;
+			const entry = Link._eventsByName.get(data[0]);
+			if (entry && entry.Event.permission) {
+				if (typeof entry.Event.permission === "string") {
+					user.checkPermission(entry.Event.permission);
+				} else {
+					entry.Event.permission(user, message);
+				}
+			}
+		}
+	}
+
+	constructor(
+		public eventName: string,
+		public allChannels: boolean,
+		public channels: Array<string | number> = [],
+		public lastRequestTime: number = 0,
+	) {
+		if (!Link._eventsByName.has(eventName)) {
+			throw new Error(`Unregistered Event class ${eventName}`);
+		}
+	}
+
+	static jsonSchema = Type.Tuple([
+		Type.String(),
+		Type.Boolean(),
+		Type.Array(Type.Union([Type.String(), Type.Number()])),
+		Type.Number(),
+	])
+
+	toJSON() {
+		return [this.eventName, this.allChannels, this.channels, this.lastRequestTime];
+	}
+
+	static fromJSON(json: Static<typeof SubscriptionRequest.jsonSchema>): SubscriptionRequest {
+		return new this(...json);
+	}
+}
+
+/**
+ * A class component to handle incoming subscription requests and offers a method to broadcast events to subscribers
+ * After creation, no other handler can be registered for SubscriptionRequest on the controller
+ */
+export class SubscriptionController {
+	_events = new Map<string, {
+		subscriptionUpdate?: SubscriptionRequestHandler<unknown>,
+		subscriptions: Map<Link, { all: boolean, channels: Array<string | number>, onceClose: () => void }>,
+	}>();
+
+	constructor(
+		private controller: any, // Controller | Link
+	) {
+		this.controller.handle(SubscriptionRequest, this._handleRequest.bind(this));
+	}
+
+	/**
+	 * Allow clients to subscribe to an event by telling the subscription controller to accept them
+	 * Has an optional subscription update handler which is called when any client updates their subscription
+	 */
+	handle<T>(Event: EventClass<T>, subscriptionUpdate?: SubscriptionRequestHandler<T>,): void;
+	handle(
+		Event: EventClass<unknown>,
+		subscriptionUpdate?: SubscriptionRequestHandler<unknown>,
+	) {
```

Comment:
This change should work. I used an overload because of how Link implemented handle, and I based my code heavily on Link but I can see that it would not be needed here.